libia2/threads.c

```diff
@@ -17,12 +18,41 @@ void *ia2_thread_begin(void *arg) {
   /* Free the thunk. */
   munmap(arg, sizeof(struct ia2_thread_thunk));
 
-  init_stacks();
   protect_tls();
+  init_stacks();
 
-  /* TODO: switch to compartment stack before calling `fn` */
+  /* Determine the current compartment so know which stack to use. */
+  uint32_t pkru = 0;
+  __asm__ volatile(
+      /* clang-format off */
+      "xor %%ecx,%%ecx\n"
+      IA2_RDPKRU "\n"
+      /* clang-format on */
+      : "=a"(pkru)::"ecx", "edx");
+  void **new_sp_addr = ia2_stackptr_for_pkey(pkru);
 
-  return fn(data);
+  /* Switch to the stack for this compartment, then call `fn(data)`. */
+  void *result;
+  __asm__ volatile(
+      /* clang-format off */
+      // Copy stack pointer to rdi.
+      "movq %%rsp, %%rdi\n"
+      #ifndef LIBIA2_INSECURE
```

Comment:
When a thread starts, it inherits the PKRU value of its creator, and so we need switch to the right stack based on the current compartment, which we don't know statically. In insecure mode, we don't know what compartment we're supposed to be in because there is no PKRU. If we really wanted, we could maintain an insecure compartment tracking variable and read it here, but I don't think there's a strong reason to.

This doesn't change transitions at other call gates, just the initial thread stack.

Reply:
Not in the case of new threads, as their initial stack pointer just points into the newly-allocated thread stack, which doesn't tell us anything about which compartment they should be running in.